### PR TITLE
Grpc deadlines

### PR DESCRIPTION
**Pull request by bondolo**
initial work in progress. The identified elements of this task include:
- [X] Extending protoc code generation to include deadline in ClientMetadata
- [ ] (?) Extending client factory to include default deadline
- [X] Adding “grpc-timeout” http header to request
- [X] Extending http header serialization to ensure timeout header is written first
- [X] set client timer for timeout
            - [X] cancel request on timeout
            - [X] report cancellation to app 
- [ ]  Reading “grpc-timeout” http header from request
- [ ]  server handle request cancel from client
- [ ]  set server timer for request timeout
   - [ ]  cancel request/response on timeout
   - [ ]  report cancellation to client

### CODE REVIEW

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java [hunk 1]
```diff
@@ -85,13 +90,50 @@ protected HttpHeaders payloadComplete(final HttpHeaders trailers) {
                 }
             };
 
+    /**
+     * GRPC spec requires timeout value to be 8 or fewer integer digits.
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java:94]: Consider linking where is this defined in the spec.

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java [hunk 2]
```diff
@@ -85,13 +90,50 @@ protected HttpHeaders payloadComplete(final HttpHeaders trailers) {
...
         assert POST.equals(request.method());
         final HttpHeaders headers = request.headers();
+        if (null != deadline && MAX_TIMEOUT.compareTo(deadline) >= 0) {
+            long timeout = deadline.toNanos(); // cannot overflow as we have already checked against safe maximum
+            int units = 0;
+            while (timeout > EIGHT_NINES) {
+                timeout = CONVERTERS[units].applyAsLong(timeout);
+                units++;
+            }
+            headers.set(GRPC_TIMEOUT_KEY, Long.toString(timeout) + ' ' + TIMEOUT_UNITS[units]);
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java:135]: There should be no space in between

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientMetadata.java [hunk 3]
```diff
@@ -42,4 +43,15 @@
      * <a href="https://www.grpc.io">gRPC</a> method.
      */
     ContentCodec requestEncoding();
+
+    /**
+     * Deadline after which the client no longer wants response.
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientMetadata.java:48]: Consider linking to https://grpc.io/blog/deadlines/

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientMetadata.java [hunk 4]
```diff
@@ -42,4 +43,15 @@
...
+     *
+     * @return {@link Duration} of associated deadline or null if no deadline. All durations greater than 99999999 hours
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientMetadata.java:50]: Consider using `{@code ...}` for `null` and `99999999`

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/DefaultGrpcClientMetadata.java [hunk 5]
```diff
@@ -31,37 +32,94 @@
...
+     */
     protected DefaultGrpcClientMetadata(final String path,
                                         @Nullable final GrpcExecutionStrategy strategy,
                                         final ContentCodec requestEncoding) {
+        this(path, strategy, requestEncoding, null);
+    }
+
+    /**
+     * Creates a new instance which uses the provided path, content codec and deadline and the default execution
+     * strategy.
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/DefaultGrpcClientMetadata.java:106]: Hard to keep consistency and correctness between `@param` and initial line when there are many overloads and code changes over time. For example, this method doesn't use "default execution strategy", it takes the execution strategy from the user. So, the javadoc is already not consistent with the params. Consider reverting back to simple "Creates a new instance". Params give reasonable explanation about the differences between overloads.

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/DefaultGrpcClientMetadata.java [hunk 6]
```diff
@@ -31,37 +32,94 @@
...
+     * @param strategy {@link GrpcExecutionStrategy} to use for the associated <a href="https://www.grpc.io">gRPC</a>
+     * method.
+     * @param requestEncoding {@link ContentCodec} to use for the associated <a href="https://www.grpc.io">gRPC</a>
+     * method.
+     * @param deadline A timeout deadline after which the response is no longer wanted.
+     */
+    protected DefaultGrpcClientMetadata(final String path,
+                                        @Nullable final GrpcExecutionStrategy strategy,
+                                        final ContentCodec requestEncoding,
+                                        @Nullable final Duration deadline) {
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/DefaultGrpcClientMetadata.java:118]: Hm, we probably should think about a builder for meta-data. The list of params keeps growing. No necessary to do that as part of this PR, but worth considering.

**Review by idelpivnitskiy — COMMENTED** [review]
Great start!
Only minor comments so far:

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientMetadata.java (hunk 4 above)
Comment (bondolo) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientMetadata.java:50]: Yes for the constant, no for null. Stephen Colebourne's JSR310 Time style Javadoc has become somewhat of defacto standard and it doesn't use `{@code }` for null, true and false. https://blog.joda.org/2012/11/javadoc-coding-standards.html

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/DefaultGrpcClientMetadata.java (hunk 5 above)
Comment (bondolo) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/DefaultGrpcClientMetadata.java:106]: fair point.

**Comment by bondolo** [comment]
Closing until this is more complete

**Pull request closed without merging**